autogen/src/header.rs

```diff
@@ -43,9 +44,55 @@ fn bit_enum_attribute() -> TokenStream {
     }
 }
 
-fn from_primitive_impl(from_prim: &[TokenStream], kind: &proc_macro2::Ident) -> TokenStream {
+fn generate_enum(
+    enum_name: &proc_macro2::Ident,
+    variants: &[(u32, proc_macro2::Ident)],
+    comment: String,
+) -> TokenStream {
+    let mut variants = variants.to_vec();
+    variants.sort_by_key(|&(number, _)| number);
+    let enumerants = variants
+        .iter()
+        .map(|(number, name)| quote! { #name = #number });
+
+    let mut number_runs = vec![];
+    number_runs.push(variants[0].0..=variants[0].0);
+    for &(number, _) in variants.iter().skip(1) {
+        let last_run = number_runs.last_mut().unwrap();
+        if number == *last_run.end() + 1 {
+            *last_run = *last_run.start()..=number;
+        } else if number > *last_run.end() + 1 {
+            number_runs.push(number..=number);
+        } else {
+            unreachable!("Variants not sorted by discriminant");
+        }
+    }
```

Comment:
```suggestion
    let mut number_runs = vec![(variants[0].0, variants[0].0)];
    for &(number, _) in variants.iter().skip(1) {
        let last_run = number_runs.last_mut().unwrap();
        if number > last_run.1 + 1 {
            number_runs.push((number, number));
        } else if number == last_run.1 + 1 {
            last_run.1 = number;
        } else {
            unreachable!("Variants not sorted by discriminant")
        }
    }
```

Using a tuple simplifies this all, and we can just use `.map(|(start, end)| {` below :)

Reply:
Done